daffodil-runtime1/src/main/scala/org/apache/daffodil/dpath/DPathRuntime.scala

```diff
@@ -311,3 +311,18 @@ abstract class Converter extends RecipeOp {
 trait ToString extends Converter {
   override def computeValue(a: DataValuePrimitive, dstate: DState): DataValueString = a.getAnyRef.toString
 }
+
+
+/**
+ * In some cases, expressions that are never used lack context and so cannot be compiled.
+ * Because they aren't used, it it safe to ignore them, but parts of Daffodil code still
+ * require a CompiledExpression. In these cases, we can use this RuntimeAbort recipe in
+ * the CompiledExpression. This recipe always throws an IllegalStateException, which
+ * treats the expression as non-constant during compilation but causes an error if
+ * evaluated at runtime.
+ */
+case class RuntimeAbortOp(expr: String) extends RecipeOp {
+  override def run(dstate: DState): Unit = {
+    throw new java.lang.IllegalStateException(s"Expression should not have been evaluated during runtime: $expr")
```

Comment:
I would expect this to be an Assert.invariantFailed(...).  

I guess I don't really understand any reason why this should get evaluated that isn't a Daffodil bug.

Reply:
After compilation we attempt to evaluate expressions to determine if they are constant or not. Throwing an `IllegalStateException` is one way that we determines it's not constant. Maybe we should be more explicit about it and do something like this:

```scala
dstate match {
  case _: DStateForConstantFolding => throw new java.lang.IllegalStateException(...)
  case _ => Abort.invariantFailed(...)
}
```

This way when we're evaluating for constant folding we throw a an excetion saying this is isn't a constant, otherwise we assume we're evaluating during a parse/unparse and that creates a more appropriate invariant?

Reply:
I've added this change to make it a bit more clear